storage/s3.go

```diff
@@ -0,0 +1,160 @@
+package storage
+
+import (
+	"bytes"
+	"fmt"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3manager"
+	"github.com/rapid7/pdf-renderer/cfg"
+)
+
+type s3Client struct {
+	client     *s3.S3
+	uploader   *s3manager.Uploader
+	downloader *s3manager.Downloader
+	bucket     string
+}
+
+// Variables needed to enforce singleton pattern for S3Client
+var (
+	client *s3Client
+	s3err  error
+)
+
+// singleton client
+func getS3Client() (*s3Client, error) {
+	if client == nil {
+		client, s3err = newS3Client()
+		if s3err != nil {
+			return nil, s3err
+		}
+	}
+	return client, nil
+}
+
+func newS3Client() (*s3Client, error) {
+	sess := session.Must(session.NewSessionWithOptions(session.Options{
+		SharedConfigState: session.SharedConfigEnable,
+	}))
+	svc := s3.New(sess)
+
+	// Attempt to create bucket
+	bucketInput := &s3.CreateBucketInput{
+		Bucket: aws.String(cfg.Config().S3Bucket()),
+	}
+
+	_, err := svc.CreateBucket(bucketInput)
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			switch aerr.Code() {
+			case s3.ErrCodeBucketAlreadyExists:
+				fmt.Printf("%s bucket already exits", cfg.Config().S3Bucket())
```

Comment:
Just noticed. Please use the logger for all these print outs.

Example:
```log "github.com/sirupsen/logrus"```
```
		log.Debug(
			fmt.Sprintf(
				"Alloc = %v MiB, TotalAlloc = %v MiB, Sys = %v MiB, NumGC = %v",
				bToMb(m.Alloc),
				bToMb(m.TotalAlloc),
				bToMb(m.Sys),
				m.NumGC,
			),
		)
```

Reply:
Out of curiosity,  does `fmt.Sprintf` inside of the `log.Debug` have any benefit over simply using `log.Debugf`?

Reply:
Not as far as I know. Feel free to use `log.Debugf`. I didn't have any specific reason for doing it this way.

Reply:
No worries, I'll keep the `fmt.Sprintf` for consistency.